### ISSUE TITLE
functools.partial of dataclass

### ISSUE THREAD

**Issue by maurosilber** [microsoft/pyright#1009]
**Is your feature request related to a problem? Please describe.**
I'm developing a framework which makes use of dataclasses to remove as much boilerplate code as possible from the user. These classes shoud use the dataclass decorator with `eq=False`.

**Describe the solution you'd like**
I'd like to be able to provide a custom decorator, which would be `my_dataclass = functools.partial(dataclass, eq=False)` or similar. But PyRight doesn't recognize

```python
@my_dataclass
class MyClass:
    x: int

MyClass(1)  # No signature completition tip
```

as a dataclass, and doesn't show signature completion tips.

Is it currently possible to do this somehow?

Thank you!

**Comment by erictraut** [comment]
Sorry, that won't work.

There are a few problems here. First, `dataclass` requires some special-case logic within Pyright, and this logic is triggered by providing a direct `@dataclass` decorator to the class. It has knowledge of specific parameter names for the `dataclass` function, and it needs to statically evaluate the arguments that are passed to these parameters. It's already very complex — more so than I like, but this complexity was needed to support the dataclass spec in a static type checker.

Second, the behavior of functools.partial cannot be described using the Python type annotation features. Once you run a function through functools.partial, almost all information about it is lost to the type checker other than its return type.

Could you have developers use `@dataclass(eq=False)` rather than wrapping this in `functools.partial`? I can't think of any other workarounds.

**Comment by LunarLanding** [comment]
With the advent of https://peps.python.org/pep-0681/ and ParamSpec, this should be addressable
